src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -3976,32 +3975,39 @@ public void reconnect(final boolean shouldRequestVersion) throws IOException, SQ
 		}
 	}
 
-	private void clientHandshakeFinalize() throws SQLException, Exception {
+	// FIXME
+	// Blocking IO
+	private CompletionStage<Void> clientHandshakeFinalize() throws SQLException, Exception {
 		// Set the default schema on the connection if one has not already been set
 		resetSchema().ifPresent(schema -> {
 			try
 			{
+				// Blocking IO
+				// Connection superclass requires setSchema to return void
+				// so CompletionStage<Void> cannot be used here
 				setSchema(schema);
```

Comment:
I'm glossing over the exceptional paths (sigh)

```suggestion
		return resetSchema()
		    .map(this::setSchemaAsync)
		    .orElseGet(StageSupport::voidStage)
		    .thenCompose(schemaUpdated -> {
		        if (setPso.isPresent()) {
 			        if (setPso.getAsLong() == -1)
 			        {
 				        // We have to turn it off
 				        return setPSO(false);
 			        }
 			        else if (setPso.getAsLong() > 0)
 			        {
 				        // Set non-default threshold
 				        return setPSO(setPso.getAsLong());
 			        }
 		        }
 		        
 		        // nothing more to do
 		        return StageSupport.voidStage();
		    });
```